module/extension/Director/src/Director.hpp

```diff
@@ -117,6 +117,103 @@ namespace module::extension {
          */
         void reevaluate_queue(provider::ProviderGroup& group);
 
+        /**
+         * An object which holds the possible solutions to running a task.
+         * Each solution represents how a single task could be executed. As each provider group can have more than one
+         * provider, there can be more than one option for how to run the task. For each option that we can choose, they
+         * can themselves have their own requirements for provider groups they need to execute. Therefore each of these
+         * requirements is themselves a solution to another task that could be emitted.
+         */
+        struct Solution {
+
+            /**
+             * A single option for how a provider could be executed
+             */
+            struct Option {
+                /// The state of this provider if it is blocked along with a description of why it is blocked
+                enum State {
+                    /// This option is not blocked to us
+                    OK,
+                    /// This provider is blocked as the task given didn't have enough authority to run it
+                    BLOCKED_PRIORITY,
+                    /// This provider is blocked due to a when condition that cannot be met
+                    BLOCKED_WHEN,
+                    /// This provider is blocked as executing it would cause a loop of providers
+                    BLOCKED_LOOP,
+                };
```

Comment:
Honestly, I'll be using `state != OK` most of the time. I put in these enums because there are differing reasons why it was blocked and when debugging it could help to be able to see the exact reason